module/mempool/stdmap/backdata/arraylinkedlist/entitiesLinkedList.go

```diff
@@ -0,0 +1,350 @@
+package arraylinkedlist
+
+import (
+	"math/rand"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type EjectionMode string
+
+const (
+	RandomEjection = EjectionMode("random-ejection")
+	LRUEjection    = EjectionMode("lru-ejection")
+)
+
+// number of entries we try to eject before giving up and ejecting LRU
+const maximumRandomTrials = 10
+
+// EIndex is data type representing an entity index in EntityDoubleLinkedList.
+type EIndex uint32
+
+// doubleLinkedListPointer represents a slice-based linked list pointer. Instead of pointing
+// to a memory address, this pointer points to a slice index.
+//
+// Note: an "undefined" (i.e., nil) notion for this doubleLinkedListPointer corresponds to the
+// value of uint32(0). Hence, legit "pointer" values start from uint32(1).
```

Comment:
Instead of implementing an undefined 'index' it like this, why not considering doing it differently? As well as `head` and `tail`, also have `size` -- the number of items in the linked list -- and assume that if `size` is > 0 then any index value will be a legit index into the array? Including a zero index...

This would also mean we can optimize away `isUndefined()` and `setUndefined()` etc. Also things like `sliceIndex()` would not have to alter the index via `-1` to account for the special nil value...

Reply:
My initial approach was also to go with `size` instead of having an `undefined` notion for `poolIndex`. However, it made the implementation even more intricate and harder to test, and hence I took the `undefined` notion path for implementation. Nevertheless, as refactoring to `size` causes a large surface of change especially on test scenarios, I issued up to your suggestion in https://github.com/dapperlabs/flow-go/issues/6082 to assess it separately. Not indenting to debate around, but the followings are among the primary reasons that I switched to `undefined` notion instead of relying on the `size`. Not saying we cannot achieve the same with the `size`, but it takes more engineering efforts and delicacies. 

1. In the current implementation, the `size` variable determines the total number of entities that are _correctly_ maintained by the pool. The implication is `size` only updated upon a successful state transition of the pool, i.e., [deceremented upon a successful invalidation](https://github.com/onflow/flow-go/blob/yahya/6014-gc-friendly-lru/module/mempool/stdmap/backdata/heropool/pool.go#L268) and [incremented upon a successful write](https://github.com/onflow/flow-go/blob/yahya/6014-gc-friendly-lru/module/mempool/stdmap/backdata/heropool/pool.go#L89). This entails a _happen before_ relationship that any successful update (i.e., invalidation or write) _must happen before_ updating the size. However, the suggested approach may break this causality. For example, right in the middle of a write, [when we claim a free head](https://github.com/onflow/flow-go/blob/yahya/6014-gc-friendly-lru/module/mempool/stdmap/backdata/heropool/pool.go#L115), we need to update the size to signal that the `free` state has a valid head index. This adds ambiguity to the implementation that the size of the pool gets updated only by claiming a free head _while we have not yet actually written anything to it_. In other words, just calling [`claimFreeHead()`](https://github.com/onflow/flow-go/blob/yahya/6014-gc-friendly-lru/module/mempool/stdmap/backdata/heropool/pool.go#L115) increments the size. Hence, any extension or development must be done very carefully to track the value of `size` and make correct assumptions about it based on the order of intermediary methods getting invoked. 

2. Having `undefined` index provides the benefit of demarcating the underlying linked-list, i.e., the `head.prev` and `tail.next` must be `undefined` which corresponds to the `nil` value in the pointer-based implementation. Demarcating the linked-list properly improves the safety and reliability of implementation. Currently, we have several tests checking the integrity of linked-list through its boundaries which helps us spot bugs much easier than the earlier version which was relying on the size. Moreover, it enables us to safely traverse the linked-list by starting at `head`, which automatically stalls at the `tail` as the `tail.next` is undefined (and similarly reverse traversal at `tail`, which automatically stalls at the `head` as the `head.prev` is `undefined`).